File: kata/Euler/isPandigal.js
class PandigitalProducts {
  constructor() {
    this.uniqueProducts = new Set();
    this.pandigitalMasks = new Array(10);

    // Pre-compute pandigital masks
    for (let n = 1; n <= 9; ++n) {
      this.pandigitalMasks[n] = (1 << (n + 1)) - 2;
    }
  }

  isPandigital(num, n) {
    let mask = 0;
    while (num > 0) {
      const digit = num % 10;
      if (digit === 0 || digit > n || (mask & (1 << digit)) !== 0) return false;
      mask |= 1 << digit;
      num = Math.floor(num / 10);
    }
    return mask === this.pandigitalMasks[n];
  }

  solve(n) {
    if (n < 2 || n > 9) {
      throw new Error("n must be between 2 and 9 inclusive");
    }

    this.uniqueProducts.clear(); // Clear any previous results

    const maxProduct = 10 ** n - 1;
    const maxMultiplicand = Math.floor(Math.sqrt(maxProduct));

    for (let i = 2; i <= maxMultiplicand; i++) {
      for (let j = i; j <= Math.floor(maxProduct / i); j++) {
        const product = i * j;
        if (product > maxProduct) break;

        const combined = `${i}${j}${product}`;
        if (combined.length !== n) continue;

        if (this.isPandigital(parseInt(combined, 10), n)) {
          this.uniqueProducts.add(product);
        }
      }
    }

    return Array.from(this.uniqueProducts).reduce((a, b) => a + b, 0);
  }
}

// Usage
const solver = new PandigitalProducts();

for (let n = 2; n <= 9; n++) {
  console.time(`Execution Time for n=${n}`);
  const result = solver.solve(n);
  console.timeEnd(`Execution Time for n=${n}`);
  console.log(`The sum of all ${n}-digit pandigital products is: ${result}`);
}

File: kata/Euler/isPandigal.cpp
#include <iostream>
#include <unordered_set>
#include <string>
#include <cmath>

using namespace std;

class PandigitalProducts {
private:
    unordered_set<int> uniqueProducts;
    int pandigitalMasks[10];

    bool isPandigital(long long num, int n) {
        int mask = 0;
        while (num > 0) {
            int digit = num % 10;
            if (digit == 0 || digit > n || (mask & (1 << digit))) return false;
            mask |= (1 << digit);
            num /= 10;
        }
        return mask == pandigitalMasks[n];
    }

public:
    PandigitalProducts() {
        // Pre-compute pandigital masks
        for (int n = 1; n <= 9; ++n) {
            pandigitalMasks[n] = (1 << (n + 1)) - 2;
        }
    }

    long long solve() {
        for (int n = 2; n <= 9; ++n) {
            int maxProduct = pow(10, n) - 1;
            
            for (int i = 1; i <= sqrt(maxProduct); ++i) {
                for (int j = i; j <= maxProduct / i; ++j) {
                    int product = i * j;
                    if (product > maxProduct) break;
                    
                    string combined = to_string(i) + to_string(j) + to_string(product);
                    if (combined.length() == n && isPandigital(stoll(combined), n)) {
                        uniqueProducts.insert(product);
                    }
                }
            }
        }

        long long sum = 0;
        for (int product : uniqueProducts) {
            sum += product;
        }
        return sum;
    }
};

int main() {
    PandigitalProducts solver;
    long long result = solver.solve();
    cout << "The sum of all pandigital products is: " << result << endl;
    return 0;
}
